src/Apps/Auction/Routes/__tests__/Register.test.tsx

```diff
@@ -0,0 +1,189 @@
+        "Address is required",
+        "Country is required",
+        "City is required",
+        "State is required",
+        "Postal code is required",
+        "Telephone is required",
+        "You must agree to the Conditions of Sale",
+      ],
+      sale_id: RegisterQueryResponseFixture.sale._id,
+      user_id: RegisterQueryResponseFixture.me.id,
+    })
+    expect(mockPostEvent).toHaveBeenCalledTimes(1)
+
+    expect(env.mutations.mockFetch).not.toBeCalled()
+    expect(window.location.assign).not.toBeCalled()
+  })
+
+  it("successfully adds a credit card and registers the user as a bidder", async () => {
+    const env = setupTestEnv()
+    const page = await env.buildPage()
+
+    setupCreateTokenMock()
+
+    env.mutations.useResultsOnce(createCreditCardSuccessful)
+    env.mutations.useResultsOnce(createBidderSuccessful)
+
+    await page.fillFormWithValidValues()
+    await page.submitForm()
+
+    expect(mockPostEvent).toBeCalledWith({
+      action_type: Schema.ActionType.RegistrationSubmitted,
+      context_page: Schema.PageName.AuctionRegistrationPage,
+      auction_slug: RegisterQueryResponseFixture.sale.id,
+      auction_state: RegisterQueryResponseFixture.sale.status,
+      bidder_id: createBidderSuccessful.createBidder.bidder.id,
+      sale_id: RegisterQueryResponseFixture.sale._id,
+      user_id: RegisterQueryResponseFixture.me.id,
+    })
+    expect(mockPostEvent).toHaveBeenCalledTimes(1)
+
+    expect(window.location.assign).toHaveBeenCalledWith(
+      `https://example.com/auction/${
+        RegisterQueryResponseFixture.sale.id
+      }/confirm-registration`
+    )
+  })
+
+  it("displays an error modal if the `createCreditCard` mutation fails", async () => {
+    const env = setupTestEnv()
+    const page = await env.buildPage()
+
+    setupCreateTokenMock()
+
+    env.mutations.useResultsOnce(createCreditCardFailed)
+
+    await page.fillFormWithValidValues()
+    await page.submitForm()
+
+    let errorModal = expectOne(page.find(ErrorModal))
+    expect(errorModal.props().show).toBe(true)
+
+    expectOne(errorModal.find(ModalButton)).simulate("click")
+
+    errorModal = expectOne(page.find(ErrorModal))
+    expect(errorModal.props().show).toBe(false)
+
+    expect(mockPostEvent).toBeCalledWith({
+      action_type: Schema.ActionType.RegistrationSubmitFailed,
+      context_page: Schema.PageName.AuctionRegistrationPage,
+      auction_slug: RegisterQueryResponseFixture.sale.id,
+      auction_state: RegisterQueryResponseFixture.sale.status,
+      error_messages: ["The `createCreditCard` mutation failed."],
+      sale_id: RegisterQueryResponseFixture.sale._id,
+      user_id: RegisterQueryResponseFixture.me.id,
+    })
+    expect(mockPostEvent).toHaveBeenCalledTimes(1)
+
+    expect(window.location.assign).not.toBeCalled()
+  })
+
+  it("displays an error modal if the `createBidder` mutation fails", async () => {
+    pending("until we can mimic Gravity-provided `createBidder` errors")
```

Comment:
Interesting. will this change in the future? (thinking about metaphysics + relay upgrades specifically)

Reply:
Yep. @ds300 and I have been discussing API changes to our `createTestEnv` tool which would support this. Another option would be to introduce a union type from Metaphysics so that the `createBidder` mutation could support a bidder _or_ an error within its response.

Here's the Slack conversation :lock: : https://artsy.slack.com/archives/C2TQ4PT8R/p1568149955022700